pgjdbc/src/main/java/org/postgresql/jdbc/TypeInfoCache.java

```diff
@@ -181,6 +182,57 @@ public int getSQLType(int oid) throws SQLException {
     return getSQLType(getPGType(oid));
   }
 
+  public void preCacheSQLTypes() throws SQLException {
+    if (cacheTypeInfoStatement == null) {
+      String sql = "SELECT typname, typinput='array_in'::regproc, typtype "
+        + "  FROM pg_catalog.pg_type "
+        + "  LEFT "
+        + "  JOIN (select ns.oid as nspoid, ns.nspname, r.r "
+        + "          from pg_namespace as ns "
+        // -- go with older way of unnesting array to be compatible with 8.0
+        + "          join ( select s.r, (current_schemas(false))[s.r] as nspname "
+        + "                   from generate_series(1, array_upper(current_schemas(false), 1)) as s(r) ) as r "
+        + "         using ( nspname ) "
+        + "         where ns.nspname  != 'pg_toast'"
+        + "       ) as sp "
+        + "    ON sp.nspoid = typnamespace "
+        + " WHERE typrelid = 0 OR (SELECT c.relkind = 'c' FROM pg_catalog.pg_class c WHERE c.oid = typrelid)"
+        + " ORDER BY typname, sp.r, pg_type.oid DESC;";
+
```

Comment:
The above query is a combination of the original query from `getSQLType(String)` plus the "where" conditions from `getTypeInfo()`

Reply:
So this is the query from psql
```
SELECT n.nspname as "Schema",
  pg_catalog.format_type(t.oid, NULL) AS "Name",
  pg_catalog.obj_description(t.oid, 'pg_type') as "Description"
FROM pg_catalog.pg_type t
     LEFT JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
WHERE (t.typrelid = 0 OR (SELECT c.relkind = 'c' FROM pg_catalog.pg_class c WHERE c.oid = t.typrelid))
  AND NOT EXISTS(SELECT 1 FROM pg_catalog.pg_type el WHERE el.oid = t.typelem AND el.typarray = t.oid)
  AND pg_catalog.pg_type_is_visible(t.oid)
ORDER BY 1, 2;
```

The one thing that is missing from ours is the is_visible which is probably a useful addition.